### DISCUSSION TITLE
CC: Tweaked 1.112.0 broke the ability to create custom computers

### DISCUSSION THREAD

**Discussion by rabiesland** [cc-tweaked/CC-Tweaked#1935] [General]
Hi, I own a mod named Embedded Computer which adds a headless computer. While I was trying to update it to [fix a crash that was reported](https://github.com/Lumina-CC/EmbeddedComputerMod/issues/1) and to add a new feature i ran into a massive issue. Things i heavily relied on had been changed. I can no longer get my peripheral API to work since it requires calling `this` in the ServerComputer constructor with the new system and to top that, the computer no longer connects to any peripherals whatsoever. 

Also, on top of the peripheral API not being able to be added, it wouldn't work if able to be added since it uses the old getAPIEnvironment to access the filesystem of a computer

Is there a way to fix this or am I just screwed?

**Comment by SquidDev** [comment]
> Also, on top of the peripheral API not being able to be added, it wouldn't work if able to be added since it uses the old `getAPIEnvironment` to access the filesystem of a computer

I'm assuming you're referring to [this code](https://github.com/Lumina-CC/EmbeddedComputerMod/blob/3a15610d39b28a4769bafeddb71fde0d3f879197/src/main/java/org/featherwhisker/embeddedcomputer/embedded/EmbeddedComputerPeripheral.java#L53-L58)? I don't especially like it, but I guess it makes sense inside the model of an embedded device. You can probably just use reflection to get the underlying `computer`.

> I can no longer get my peripheral API to work since it requires calling `this` in the `ServerComputer` constructor with the new system and to top that, the computer no longer connects to any peripherals whatsoever.

Sorry, I'm really confused by this section. Nothing should have changed with peripherals in this release. Looking at your code, do you actually mean custom APIs? Looking at your API, I don't really see much benefit to it — that probably could just be a built-in program and/or module. However, you can probably do some thing where you have a "computer holder" class, pass that as a custom component, and then tie the knot afterwards.

**Reply by rabiesland** [reply]
Sorry, I did mean custom API. I used the ServerComputer methods to add apis

**Comment by rabiesland** [comment]
The peripheral issue seems to stem back to the embedded API not loading, which is used by autostart programs to signify that a computer is embedded and it does all the weird peripheral stuff i was using to make sure they still worked

I can probably make that API a stub and implement all the functions in lua
